File: apps/mobile/ios/Sources/FlagshipUI/ViewModels/ActivityViewModel.swift
import Foundation
import Observation
import FlagshipAPI

public struct ActivityFeed: Sendable {
    public let recentInstalls: [RecentInstallEvent]
    public let pairedSessions: [PairedSessionSummary]
    /// Non-nil when the daemon reports a J.3/J.4 reissuance snapshot
    /// (post-recovery walk). Surfaces as a card on Activity that
    /// links into the Settings → Recovery → Re-attach progress
    /// screen. Mirrors webapp activity.js fan-out.
    public let postRecovery: PostRecoverySnapshot?
    /// Account-level audit events from flagshipserver.com — device
    /// disconnects, IRK rotations, wipe-restart, recovery setup. Pulled
    /// alongside the daemon-side install events so the user sees their
    /// account history in one place. Sorted descending by seq.
    public let auditEvents: [AuditEvent]
}

@Observable
@MainActor
public final class ActivityViewModel {
    public private(set) var state: LoadingState<ActivityFeed> = .idle

    private let client: any ScreensClient
    /// Optional — only wired in production. When nil (older callers /
    /// tests that don't care about audit events), the audit section
    /// stays empty and the rest of the feed still loads.
    private let server: (any FlagshipServerClient)?
    private let currentUsername: @MainActor () -> String?

    public init(
        client: any ScreensClient,
        server: (any FlagshipServerClient)? = nil,
        username: @MainActor @escaping () -> String? = { nil }
    ) {
        self.client = client
        self.server = server
        self.currentUsername = username
    }

    public func load() async {
        state = .loading
        do {
            async let detail = client.serverDetail()
            async let sessions = client.pairedSessionsList()
            // Post-recovery is tolerated as missing — a daemon that
            // hasn't shipped P1.23 returns a non-2xx, and we render
            // the rest of the feed regardless. Mirrors the
            // try/catch in apps/web/public/webapp/views/activity.js.
            async let recovery: PostRecoverySnapshot? = {
                do { return try await client.postRecoveryStatus().report }
                catch { return nil }
            }()
            // Audit events live on flagshipserver.com (.com control
            // plane), not on the daemon. Tolerated as missing too —
            // a fresh install or a misconfigured Worker shouldn't
            // break the rest of the feed.
            // Capture deps before the async let — `currentUsername`
            // is MainActor-isolated so it has to be read on the
            // outer task, not inside the unstructured child.
            let username = currentUsername()
            let serverRef = server
            async let audit: [AuditEvent] = ActivityViewModel.fetchAuditOrEmpty(
                server: serverRef,
                username: username,
            )
            let feed = try await ActivityFeed(
                recentInstalls: detail.recentInstallEvents,
                pairedSessions: sessions.sessions,
                postRecovery: recovery,
                auditEvents: audit
            )
            state = .loaded(feed)
        } catch {
            state = .failed(error.localizedDescription)
        }
    }

    /// Helper for the audit-events async let. Static + nonisolated so
    /// it can be called from an unstructured child task without
    /// crossing actor isolation. Returns [] on any error — audit is
    /// non-essential, never breaks the main feed.
    private static func fetchAuditOrEmpty(
        server: (any FlagshipServerClient)?,
        username: String?,
    ) async -> [AuditEvent] {
        guard let server, let username, !username.isEmpty else { return [] }
        do {
            return try await server.listAuditEvents(username: username, sinceSeq: 0, limit: 20).events
        } catch {
            return []
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/viewmodels/ActivityViewModel.kt
// Aggregate "activity feed": pending unlock-approvals + recent
// install events + post-recovery snapshot, sorted by time. Mirrors
// FlagshipUI/ViewModels/ActivityViewModel.swift.

package com.flagshipserver.app.viewmodels

import com.flagshipserver.app.api.AuditEvent
import com.flagshipserver.app.api.FlagshipServerClient
import com.flagshipserver.app.api.RecentInstallEvent
import com.flagshipserver.app.api.ScreensClient
import com.flagshipserver.app.api.PostRecoverySnapshot
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

sealed interface ActivityItem {
    val at: Long
    val title: String
    val subtitle: String?

    data class InstallEvent(
        val event: RecentInstallEvent,
        override val at: Long = event.at,
        override val title: String = "${event.kind}: ${event.serviceId}",
        override val subtitle: String? = event.detail,
    ) : ActivityItem

    data class RecoverySnapshot(
        val snapshot: PostRecoverySnapshot,
        override val at: Long =
            snapshot.lastReissue?.completedAt
                ?: snapshot.lastReissue?.startedAt
                ?: 0L,
        override val title: String = "Recovery: ${snapshot.lastReissue?.status ?: "snapshot"}",
        override val subtitle: String? =
            snapshot.lastReissue?.let { "${it.totalRewritten} rewrites · ${it.reattachedCount} reattached" },
    ) : ActivityItem

    /** Account-level audit event — surfaced alongside install events
     *  in the merged time-sorted feed. eventKind matches the Worker's
     *  controlled vocabulary; subtitle carries the human-readable
     *  detail string the Worker recorded ("Disconnected iPad (kitchen)"). */
    data class AuditEntry(
        val event: AuditEvent,
        override val at: Long = event.postedAt,
        override val title: String = auditLabel(event.eventKind),
        override val subtitle: String? = event.detail.takeIf { it.isNotBlank() },
    ) : ActivityItem
}

private fun auditLabel(kind: String): String = when (kind) {
    "device-disconnected" -> "Disconnected device"
    "device-replaced"     -> "Replaced device"
    "device-added"        -> "Added device"
    "wipe-restart"        -> "Wiped & restarted account"
    "recovery-set-up"     -> "Set up recovery"
    "recovery-rotated"    -> "Rotated recovery passkey"
    "app-renamed"         -> "Renamed app URL"
    "server-created"      -> "Created server"
    "server-online"       -> "Server came online"
    else                  -> kind
}

data class ActivityFeed(
    val items: List<ActivityItem>,
)

class ActivityViewModel(
    private val client: ScreensClient,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO),
    /** Optional — only wired in production. When null, the audit
     *  section silently stays empty so older callers + tests that
     *  don't care about account-level events still work. Kept after
     *  `scope` so existing `(client, scope)` positional callers
     *  (the test fixtures) stay binding-compatible. */
    private val server: FlagshipServerClient? = null,
    /** Captured rather than stored so the VM picks up AppState
     *  changes (post sign-in / sign-out) without a re-init. */
    private val username: () -> String? = { null },
) {
    private val _state = MutableStateFlow<LoadingState<ActivityFeed>>(LoadingState.Idle)
    val state: StateFlow<LoadingState<ActivityFeed>> = _state.asStateFlow()

    fun load() = scope.launch {
        _state.value = LoadingState.Loading
        // coroutineScope contains the structured-concurrency tree so a
        // sibling async{} failure cancels its siblings cleanly and
        // bubbles out as a single throw the outer try{} catches — no
        // dangling uncaught exceptions in the parent scope.
        val outcome = runCatching {
            coroutineScope {
                val detailJob = async { runCatching { client.serverDetail().recentInstallEvents }.getOrDefault(emptyList()) }
                val recoveryJob = async { runCatching { client.postRecoveryStatus().report }.getOrNull() }
                // Account-level audit feed. Lives on .com (not the
                // daemon) so the fetch is tolerated as missing (a
                // misconfigured Worker shouldn't break the rest of
                // the feed).
                val auditJob = async {
                    val u = username()
                    val s = server
                    if (s == null || u.isNullOrEmpty()) emptyList()
                    else runCatching { s.listAuditEvents(u, sinceSeq = 0, limit = 20).events }.getOrDefault(emptyList())
                }
                Triple(
                    detailJob.await(),
                    recoveryJob.await(),
                    auditJob.await(),
                )
            }
        }
        outcome.fold(
            onSuccess = { (recents, snapshot, audit) ->
                val items = buildList {
                    recents.forEach { add(ActivityItem.InstallEvent(it)) }
                    snapshot?.let { add(ActivityItem.RecoverySnapshot(it)) }
                    audit.forEach { add(ActivityItem.AuditEntry(it)) }
                }.sortedByDescending { it.at }
                _state.value = LoadingState.Loaded(ActivityFeed(items = items))
            },
            onFailure = { t ->
                _state.value = LoadingState.Failed(t.message ?: "couldn't load activity")
            },
        )
    }
}
